src/pip/_internal/models/wheel.py

```diff
@@ -32,22 +39,27 @@ def __init__(self, filename: str) -> None:
         self.name = wheel_info.group("name").replace("_", "-")
         _version = wheel_info.group("ver")
         if "_" in _version:
-            deprecated(
-                reason=(
-                    f"Wheel filename {filename!r} uses an invalid filename format, "
-                    f"as the version part {_version!r} is not correctly normalised, "
-                    "and contains an underscore character. Future versions of pip may "
-                    "fail to recognise this wheel."
-                ),
-                replacement=(
-                    "rename the wheel to use a correctly normalised version part "
-                    "(this may require updating the version in the project metadata)"
-                ),
-                gone_in="25.1",
-                issue=12914,
-            )
             _version = _version.replace("_", "-")
 
+            try:
+                parse_wheel_filename(filename)
+            except (PackagingInvalidWheelName, InvalidVersion):
```

Comment:
By just reading the code here, it's not obvious that all wheel filename parsing errors here are guaranteed to relate to an invalid version part.

Would it make sense to catch the two exception separately with a slightly different deprecation message?

Or is there a guarantee here that all exceptions are due to the version part? In that case a little comment explaining that might be useful.

Reply:
It's not guaranteed but it would have to occur under the following series of events:

1. The wheel filename passes the existing regex
2. There is a `_` that is valid in what the regex parses as "version" but is not actually part of the version
3. There is a different error in the filename that packaging catches

This would therefore have to be a currently unknown error that wheel filenames can have that packaging does not support but pip's regex was passing on anyway.

The reason I am catching both exceptions is because I did the same in https://github.com/pypa/pip/pull/12327 and I was borrowing that code, but in that PR the code was scanning a directory that could find arbitrary names and I saw both errors in my testing.

Happy to split catching the two different exceptions up and give an appropriate message for `PackagingInvalidWheelName` .